Added 'swipe down to refresh' to Live Matches

Rhacco opened the pull request:
See #10 

Added the very common 'swipe down to refresh' action to Live Matches to allow refreshing the data without having to reopen the app entirely. A new live match is added to the list and a finished match is removed from the list by utilizing GetMatchDetails from the official Dota 2 API.

Rhacco committed:
Added 'swipe down to refresh' to Live Matches

Rhacco committed:
Fixed Live Matches being updated twice

... when e.g. orientation changed

Rhacco committed:
Prevented another bug from creating duplicates

This one is very strange but before this commit, a change in orientation
would create one single duplicate in Live Matches pretty much every
time. I don't know the cause and I think it would be very hard to debug
so now I'm just preventing the creation of duplicates entirely. Also
added a comment where it wasn't clear at all what's happening.

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -29,21 +39,26 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
-            mListAdapter.mListItemsData.add(0, listItemData)
+    private fun updateLiveMatchesListData(newItem: LiveMatchesListAdapter.ListItemData) {
+        // Strangely, sometimes the same match arrives here twice, so prevent creating duplicates
+        mListAdapter.mListItemsData
```
it looks like you're only adding on matches that haven't previously been shown and invalidating the list after you add on the new matches.

UI wise, this is very expensive. You could do the same thing with recyclerview and do a notifyItemAdded() to save on the UI rendering.

Next time you test this, look for the "Frames dropped" message on Logcat. Every 16ms, Android refreshes the screen and if there is too much work on the UI thread, it will skip that refresh. This can lead to a laggy user experience. For our intents and purposes, we shouldn't see that since this app is very simple at this point.

Also, why don't we just invalidate all the matches and just show the most recent ones? What benefit do we get  from adding only the matches not shown in terms of user experience? I kind of understand where you're going but I just want to know why.

Rhacco replied:
Currently I'm doing a full wipe and new fetch of all live matches. With the migration to RecyclerView in #9 I will use notifyItemInserted() later, instead of wiping everything

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -29,21 +39,26 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
+                .filter { it.mMatchID == newItem.mMatchID }
+                .forEach { return }
+
+        if (newItem.mAverageMMR < 1) {
```
Why is this mAverageMMR evaluation needed?

Rhacco replied:
GetTopLiveGame also returns currently running tournament matches and they will have an average MMR of 0. So this evaluation checks if it needs to handle a tournament or a ranked match (which will have average MMR > 0)

EdbertChan replied:
Please comment this observation in the code as this detail is not self-evident in the logic. In the future, we maybe put this in a helper.

Rhacco replied:
I agree, thanks. Done in #9

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -29,21 +39,26 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
             return
         }
-        mListAdapter.mListItemsData.add(listItemData)
+        mListAdapter.mListItemsData.add(newItem)
```
Please don't ever access variables directly like this. Also, google "Law of Demeter" because this is exposing the internal guts of the listAdapter when it should be used primarily to hide the data source and only interact with the list.

If anything, please just use a mAdapter.setList() and mAdapter.getList().

Rhacco replied:
Fixed in #9

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -29,21 +39,26 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
         var index = 0
         for (item in mListAdapter.mListItemsData) {
-            if (item.mAverageMMR < 1 || item.mAverageMMR >= listItemData.mAverageMMR) {
+            if (item.mAverageMMR < 1 || item.mAverageMMR >= newItem.mAverageMMR) {
```
What is the idea behind  || item.mAverageMMR >= newItem.mAverageMMR?

Rhacco replied:
Sorting by descending average MMR (see comment for method)

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -29,21 +39,26 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
-        if (listItemData.mAverageMMR < 1) {
-            mListAdapter.mListItemsData.add(0, listItemData)
+    private fun updateLiveMatchesListData(newItem: LiveMatchesListAdapter.ListItemData) {
+        // Strangely, sometimes the same match arrives here twice, so prevent creating duplicates
```
Uhhh...... "I dont know" is not a good sign this early into development. Please find the reason why.

Rhacco replied:
No longer relevant

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/viewmodel/MatchesViewModel.kt:
```diff
@@ -36,20 +36,23 @@ class MatchesViewModel(application: Application) : AndroidViewModel(application)
+                                newItem.mTitle = "Ranked Match (Average MMR " + averageMMR +
                                         ", Match ID " + result.match.matchid + ")"
                             if (result.teams.size == 2) {
                                 result.teams[0].players
```
Don't repeat yourself. You have result.team[0] and result.team[1] that can easily be put into a loop (or stream)

Also, when is result.team.size not equal to 2?

Rhacco replied:
Already fixed in #9 

We are fetching data from an external source that might malfunction or change at some point. So I'd rather add a quick safety check now and some debug information to make debugging easier when needed

Rhacco committed:
Merge branch 'feature/ShowLiveMatches-#7'

... into feature/RefreshAction-#10

Rhacco committed:
Removed unused files not recognized by merge

Rhacco committed:
Merge branch 'master'

... into feature/RefreshAction-#10

Rhacco committed:
Now adding new live matches to old ones on refresh

... instead of wiping everything

Rhacco committed:
Fixed onCreated being called on orientation change

Rhacco committed:
Now removing finished matches

... after updating live matches. GetMatchDetails from the official Dota
2 API is used to check if there are match details present for given
match ID. If yes, match is finished

Rhacco committed:
Better item insertions, updated comment

Rhacco committed:
Don't check if match finished when just added

EdbertChan commented on app/src/main/kotlin/github/com/rhacco/dota2androidapp/activities/LiveMatchesActivity.kt:
```diff
@@ -38,7 +42,20 @@ class LiveMatchesActivity : BaseLifecycleActivity<MatchesViewModel>() {
+            it?.let { (matchId, isFinished) ->
+                if (isFinished) {
+                    mAdapter.remove(matchId)
+                    mViewModel.removeFinishedMatch(matchId)
```
Having to do the deletion on both the adapter and the viewmodel defeats the purpose of the viewmodel providing the updates. 

I would suggest just calling .removeFinishedMatch and opening a channel on the viewmodel to listen to when the matchId is removed. Once that occurs, then call mAdapter.remove().

Rhacco replied:
Done, thanks

Rhacco committed:
Following the ViewModel architecture again
